src/tools/kotlin-schema-field.ts

```diff
@@ -0,0 +1,268 @@
+        case 'schema-ordered-list': return new ListType(constructType(descriptor.schema));
+        case 'schema-reference': return new ReferenceType(constructType(descriptor.schema));
+        case 'type-name': return new ReferencedSchema(node.refs.get(name));
+        case 'schema-inline': return new ReferencedSchema(node.refs.get(name));
+        case 'schema-nested': return new InlineSchema(node.refs.get(name));
+        case 'schema-primitive':
+        case 'kotlin-primitive': return new PrimitiveType(descriptor.type);
+        default:
+          throw new Error(`Schema kind '${descriptor.kind}' is not supported`);
+      }
+    })(descriptor);
+
+    return new KotlinSchemaField(name, type);
+  });
+}
+
+/** Represents a field in a schema */
+export class KotlinSchemaField {
+  constructor(readonly name: string, readonly type: SchemaFieldType) {}
+}
+
+/**
+ * Root of the type hierarchy of schema fields:
+ *
+ * SchemaFieldType
+ *  ├─TypeContainer
+ *  │  ├─CollectionType
+ *  │  ├─ListType
+ *  │  └─ReferenceType
+ *  ├─SchemaType
+ *  │  ├─ReferencedSchema
+ *  │  └─InlineSchema
+ *  └─PrimitiveType
+ */
+export abstract class SchemaFieldType {
+
+  abstract get defaultVal(): string;
+
+  get isCollection() {
+    return false;
+  }
+
+  get isSchema() {
+    return false;
+  }
+
+  get isPrimitive() {
+    return false;
+  }
+
+  get decodeFn(): string {
+    throw new Error('Decoding function not available for this type');
+  }
+
+  get schemaHash(): string {
+    throw new Error('Schema Hash not available for this type');
+  }
+
+  get arcsTypeName(): string {
+    throw new Error('Arcs Type Name not available for this type');
+  }
+
+  get kotlinType() {
+    return this.getKotlinType(false);
+  }
+
+  abstract unwrap(): SchemaFieldType;
+
+  abstract getKotlinType(contained: boolean): string;
+}
+
+/** Super class of all type containers */
+abstract class TypeContainer extends SchemaFieldType {
+
+  constructor(readonly innerType: SchemaFieldType) {
+    super();
+  }
+
+  get innerKotlinType() {
+    return this.innerType.getKotlinType(true);
+  }
+
+  unwrap() {
+    return this.innerType;
+  }
+}
+
+/** Collection type, e.g. [Number], [&Thing {...}] */
+class CollectionType extends TypeContainer {
+
+  constructor(innerType: SchemaFieldType) {
+    super(innerType);
+  }
+
+  get isCollection() {
+    return true;
+  }
+
+  getKotlinType() {
+    return `Set<${this.innerKotlinType}>`;
+  }
+
+  get defaultVal() {
+    return `emptySet<${this.innerKotlinType}>()`;
+  }
+
+  get decodeFn() {
+    return this.innerType.decodeFn;
+  }
+}
+
+/** List type, e.g. List<Text>, List<inline Object {...}> */
+class ListType extends TypeContainer {
+
+  constructor(innerType: SchemaFieldType) {
+    super(innerType);
+  }
+
+  getKotlinType() {
+    return `List<${this.innerKotlinType}>`;
+  }
+
+  get defaultVal() {
+    return `listOf<${this.innerKotlinType}>()`;
```

Comment:
`emptyList()` -- unless, do we need to specify the type information?

Reply:
I doesn't seem like we have to given the type explicitly, updating.